### ISSUE TITLE
Excessive memory usage

### ISSUE THREAD

**Issue by tomchapin** [armzilla/amazon-echo-ha-bridge#26]
When I use Java to run the JAR from Windows 10 (with JDK 1.8), everything works perfectly, but the process is using up 1.14 gigabytes of RAM.

Is this intended behavior? Is there anything I can do to run the server so it doesn't use up so much RAM?

P.S. Thanks for this awesome tool!

**Comment by aptalca** [comment]
I noticed that, too. I am running it as a docker container with an ubuntu base and yup, it is using between 850MB-1.1GB at all times.

Since so many people are running this on a raspberry pi, I am curious as to whether this is their experience as well (rpi 2 only has 1GB of ram)

Also looking at the logs, I see that this app remains in discovery mode forever and it gets absolutely hammered with discovery requests from devices on my network (new request every ten seconds). I wonder if that has anything to do with the memory usage

**Comment by jlekhter** [comment]
I'm running this on a rpi 2 -- its been up for 22 days now and top shows 139m resident memory and 362m virtual memory.

I agree -- awesome software. Thank you!

**Comment by armzilla** [comment]
I think that is default java behavior to adjust usage based on your total memory size.  You try and specify the  -Xmx512M flag.  e.g. java -Xmx512M -jar <bridge jar>.

**Comment by aptalca** [comment]
Thanks, Arm. That was it. I set it to 512M and now the app itself consumes under 500MB.

Is 512MB a reasonable max value or could we go lower? I am running this on a server that does a whole lot of tasks so I would rather not have java tie up precious resources when not needed.

Thanks

**Comment by armzilla** [comment]
I think the majority of the memory usage is the embedded elasticsearch.  I would imagine you can drop it down to 256M and see if it causes any issues.

**Comment by maddox** [comment]
Why in the world does this project use elasticsearch at all anyways? It takes requests on unique specific URLs and then makes requests to the URLs mapped to those. It's all very weird to me.

**Comment by armzilla** [comment]
Its embeddable, didn't want folks to have to set up mongo and such.  Didn't want to set up relational db because I want to be able to upgrade the data stored from one version to next (without having to think about migration strategy).  Guess I could have stored things as clobs in rdb.

**Comment by maddox** [comment]
You could honestly just serialize it to JSON on disk ¯\(°_o)/¯ It's not a ton of data. Keep it in memory, and dump it to a JSON file. I assume the only information being stored are the "devices" we create? 

It'd get rid of a ton of dependencies as well as making the app way lighter. And seeing that people run this on Raspberry Pis, that'd be a huge benefit.

**Comment by armzilla** [comment]
Guess I could have gone with the embedded derby route as well.  Hindsight 20/20 lol

**Comment by armzilla** [comment]
Yep, I run mine on a pi2 with no problems.  I guess the whole es route cam from my other HA project to also store   sensor events.
